packages/zoe/src/contracts/multipoolAutoswap/swap.js

```diff
@@ -22,95 +34,95 @@ export const makeMakeSwapInvitation = (
     });
     const {
       give: { In: amountIn },
-      want: { Out: wantedAmountOut },
+      want: {
+        Out: { brand: brandOut },
+      },
     } = seat.getProposal();
-    const { brand: brandIn, value: inputValue } = amountIn;
-    const brandOut = wantedAmountOut.brand;
+    const { brand: brandIn } = amountIn;
 
     // we could be swapping (1) central to secondary, (2) secondary to
     // central, or (3) secondary to secondary
 
     if (isCentral(brandIn) && isSecondary(brandOut)) {
       const pool = getPool(brandOut);
-      const amountOut = pool.getCentralToSecondaryInputPrice(inputValue);
+      const {
+        amountOut,
+        amountIn: reducedAmountIn,
+      } = pool.getPriceGivenAvailableInput(amountIn, brandOut);
       trade(
         zcf,
         {
           seat: pool.getPoolSeat(),
-          gains: { Central: amountIn },
+          gains: { Central: reducedAmountIn },
           losses: { Secondary: amountOut },
         },
         {
           seat,
           gains: { Out: amountOut },
-          losses: { In: amountIn },
+          losses: { In: reducedAmountIn },
         },
       );
       seat.exit();
       return `Swap successfully completed.`;
-    }
-
-    if (isSecondary(brandIn) && isCentral(brandOut)) {
+    } else if (isSecondary(brandIn) && isCentral(brandOut)) {
+      // this branch is very similar to the above, with only pool and the left
+      // seat's gains and losses changing. Sharing code makes it less readable.
       const pool = getPool(brandIn);
-      const amountOut = pool.getSecondaryToCentralInputPrice(inputValue);
+      const {
+        amountOut,
+        amountIn: reducedAmountIn,
+      } = pool.getPriceGivenAvailableInput(amountIn, brandOut);
       trade(
         zcf,
         {
           seat: pool.getPoolSeat(),
-          gains: { Secondary: amountIn },
+          gains: { Secondary: reducedAmountIn },
           losses: { Central: amountOut },
         },
         {
           seat,
           gains: { Out: amountOut },
-          losses: { In: amountIn },
+          losses: { In: reducedAmountIn },
         },
       );
       seat.exit();
       return `Swap successfully completed.`;
-    }
-
-    if (isSecondary(brandIn) && isSecondary(brandOut)) {
-      // We must do two consecutive `getCurrentPrice` calls: from
-      // the brandIn to the central token, then from the central
-      // token to the brandOut.
+    } else if (isSecondary(brandIn) && isSecondary(brandOut)) {
```

Comment:
architecturally, all these giant `if` statements need to go. That's what OO is for

Reply:
agreed. There's an issue for that.